### PR TITLE
Refactored authorizers to classes

### PR DESCRIPTION
**Pull request by mcq8**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

The idea of this PR is to refactor all the authentication to classes and have less global methods.
It should now be more clear what to implement if you want to add an authoriser.

### CODE REVIEW

**Comment by mcq8** [comment]
I forgot about php 5.3, I'm just to used to shortened array syntax

**Comment by murrant** [comment]
Thanks for working on this.  Didn't have time to give this a full look yet, but here are some initial thoughts.

You can avoid the global variable $authorizer by making a proxy with static methods for the singleton authorizer.  You could also put your factory in that class and use the autoloader instead of sourcing a php file.

Please use Config::get() instead of direct access to $config, we are trying to remove global variables :)

Can move the interface to `LibreNMS\Interfaces\`?

File: html/pages/edituser.inc.php [hunk 1]
```diff
@@ -249,10 +251,10 @@
                     $vars['can_modify_passwd'] = '1';
                 }
 
-                $authorizer->updateUser($vars['user_id'], $vars['new_realname'], $vars['new_level'], $vars['can_modify_passwd'], $vars['new_email']);
+                AuthorizerFactory::get()->updateUser($vars['user_id'], $vars['new_realname'], $vars['new_level'], $vars['can_modify_passwd'], $vars['new_email']);
                 print_message('User has been updated');
-                if (!empty($vars['new_pass1']) && $vars['new_pass1'] == $vars['new_pass2'] && $authorizer->passwordscanchange($vars['cur_username'])) {
-                    if ($authorizer->changepassword($vars['cur_username'], $vars['new_pass1']) == 1) {
+                if (!empty($vars['new_pass1']) && $vars['new_pass1'] == $vars['new_pass2'] && \LibreNMS\Authentication\AuthorizerFactory::get()->passwordscanchange($vars['cur_username'])) {
```
Comment (murrant) [inline html/pages/edituser.inc.php]: oops unneeded full namespace here.

File: LibreNMS/Authentication/AuthorizerFactory.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
+<?php
+
+namespace LibreNMS\Authentication;
+
+use LibreNMS\Config;
+use LibreNMS\Interfaces\Authentication\Authorizer;
+
+class AuthorizerFactory
+{
+    protected static $_instace;
```
Comment (murrant) [inline LibreNMS/Authentication/AuthorizerFactory.php]: Probably a typo here `$_instace` -> `$_instance`

**Comment by murrant** [comment]
Looks pretty good, just needs testing.  Too bad we can't use dependency injection eh?

What auth methods did you test?

**Comment by mcq8** [comment]
Still doing the Config.
Only the mysql so far I don't have an AD.

**Comment by murrant** [comment]
Good, I can test AD

File: LibreNMS/Authentication/LdapAuthorizationAuthorizer.php [hunk 3]
```diff
@@ -134,8 +134,8 @@ public function getUserlevel($username)
         // Loop the list and find the highest level
         foreach ($entries as $entry) {
             $groupname = $entry['cn'][0];
-            if (Config::get('auth_ldap_groups')[$groupname]['level'] > $userlevel) {
-                $userlevel = Config::get('auth_ldap_groups')[$groupname]['level'];
+            if (Config::get('auth_ldap_groups.' . $groupname . '.level') > $userlevel) {
```
Comment (murrant) [inline LibreNMS/Authentication/LdapAuthorizationAuthorizer.php]: You can't do this because some group names have periods in the names :)

Comment (mcq8) [reply LibreNMS/Authentication/LdapAuthorizationAuthorizer.php]: Seems like there are a couple of more I need to fix, php 5.3 does not like the array acces directly on the return of a function.
https://3v4l.org/hcOHo#output

Comment (mcq8) [reply LibreNMS/Authentication/LdapAuthorizationAuthorizer.php]: @murrant is it ok to just make a temp var for it right now?
```
            $authLdapGroups = Config::get('auth_ldap_groups');
            if ($authLdapGroups[$groupname]['level'] > $userlevel) {
                $userlevel = $authLdapGroups[$groupname]['level'];
            }
```
Not sure I want to drag new methods into those classes as part of this refactoring.
The cleanup of the ldap and AD duplication can be for an other refactoring.

Comment (murrant) [reply LibreNMS/Authentication/LdapAuthorizationAuthorizer.php]: Yes, that is the best way to fix for now.

**Comment by laf** [comment]
@mcq8 Sorry needs a rebase now due to another merged PR.

**Comment by murrant** [comment]
@mcq8 we merged an update to ldap authentication.  Are you still planning on updating this?

**Comment by mcq8** [comment]
yes I'll rebase

File: LibreNMS/Authentication/ADAuthorizationAuthorizer.php [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+    {
+        $userlist = array();
+        $userhash = array();
+
+        $ldap_groups = $this->getGroupList();
+
+        foreach ($ldap_groups as $ldap_group) {
+            $search_filter = "(memberOf=$ldap_group)";
+            if (Config::get('auth_ad_user_filter')) {
+                $search_filter = "(&{Config::get('auth_ad_user_filter')}$search_filter)";
```
Comment (murrant) [inline LibreNMS/Authentication/ADAuthorizationAuthorizer.php]: Typo here causes the Config::get to be in the filter instead of the intended output.

**Comment by murrant** [comment]
AD alerting tested.

**Comment by murrant** [comment]
Looks good to me.  Would be good to have ldap and radius tests. But they looked ok to me.

**Comment by papatango71** [comment]
The update to LibreNMS/Authentication/LdapAuthorizationAuthorizer.php  breaks my /overview pages.  I am able to log in and and see everything else except my main overview dashboard.  Below is the error that I got when running /overview/debug=yes

Array ( [0] => 1 [1] => Uncaught Error: Call to undefined method LibreNMS\Authentication\LdapAuthorizer::ldap_compare() in /opt/librenms/LibreNMS/Authentication/LdapAuthorizer.php:152 Stack trace: #0 /opt/librenms/LibreNMS/Authentication/LdapAuthorizer.php(178): LibreNMS\Authentication\LdapAuthorizer->getUserlist() #1 /opt/librenms/html/includes/functions.inc.php(1606): LibreNMS\Authentication\LdapAuthorizer->getUser('4') #2 /opt/librenms/html/pages/front/tiles.php(23): get_dashboards() #3 /opt/librenms/html/index.php(218): require('/opt/librenms/h...') #4 {main} thrown [2] => /opt/librenms/LibreNMS/Authentication/LdapAuthorizer.php [3] => 152 )

**Pull request merged**